src/main/scala/is/hail/variant/VariantView.scala

```diff
@@ -44,14 +37,15 @@ class VariantView(tv: TVariant) {
     TString.loadString(region, t.loadField(region, offset, refIdx))
   }
 
-  def getAltAlleles(): Long = {
-    t.loadField(region, offset, altAllelesIdx)
+  val altAlleles = new ArrayView(tAltAlleles, new AltAlleleView(tAltAllele))
+  def loadAltAlleles() {
```

Comment:
You don't want to just do this in `setRegion`?  Seems pretty innocuous (an add and two stores).

Reply:
Yeah, I find myself often unsure of what is and is not reasonable these days. 😓

1. This approach is more likely to result in API-user error. In exchange, you run only those instructions necessary for what you want.

2. Loading in `setRegion` eliminates API-user error. In exchange, we pay an add and two stores. This seems OK if most algorithms rely on the alt alleles anyway. Also seems OK generally, since it's not a lot of work? (not sure how to judge that claim)

3. I also considered lazy loading of the array offset as with `ref` and `alt` in `AltAlleleView`. This eliminates API-user error. You only pay the add and two stores if you want the alt alleles. However, every access to alt alleles has a conditional branch. After the second access, the condition always fails, so hopefully several accesses play well with a branch predictor (right?).

How should I be thinking about this? I tend to rely on "go with the safe thing," but, for me, that maxim has faded in our current development milieu. Maybe I should stick to it? That leaves two and three. The maxim "don't prematurely optimize" would say go for two. Does this seem right to you?

Reply:
A few rules I'm using to guide this stuff:

1. If it isn't "per element" of the aggregated data structure, don't worry about it too much.  So don't sweat per-variant stuff in VSM but do sweat per-genotype and per-row in a KT.

2. We're far from optimizing at the level of a few instructions.  Do what's easiest and don't sweat a few instructions.

3. Allocating strings here probably dominate everything else, so don't worry about deferring a load or two until we have a better story for strings.

Yes, your (2.)  seems fine to me.

Writing unsafe destructing stuff is tedious, so I see the views on the Variant, etc. objects mainly just to package that up and make the client code look nicer.